src/lib.rs

```diff
@@ -134,12 +142,15 @@ impl StdinOverride {
     /// Reset the standard input to its state before this type was constructed.
     ///
     /// This can be called to manually handle errors produced by the destructor.
-    pub fn reset(mut self) -> io::Result<()> {
+    pub fn reset(self) -> io::Result<()> {
         self.reset_inner()?;
-        self.reset = true;
+        std::mem::forget(self);
         Ok(())
     }
     fn reset_inner(&self) -> io::Result<()> {
+        if OVERRIDDEN_STDIN_COUNT.fetch_sub(1, Ordering::SeqCst) != self.index {
+            panic!("Stdin override reset out of order!");
+        }
```

Comment:
spit-balling here feel free to leave your thoughts, isn' this more correct?
```rust
if OVERRIDDEN_STDIN_COUNT.compare_and_swap(self.index, self.index-1, Ordering::SeqCst) != self.index {
....
}
```
My reasoning is what if someone tries to recover from such bug, or if he has multiple of these in different threads and this leaves the atomic in a "bad" state (the index is 1 less even though nothing was dropped)
With CAS the value is left as is if the panic was recovered.

Reply:
In that case we should just swap right? Because if we compare and swap then it doesn't change the value if it isn't `self.index`.

Reply:
Why should it swap the value if it isn't `self.index`?

Reply:
My reasoning is:
If you failed to drop it correctly the value shouldn't change, that way you could still potentially drop other things correctly.
if you swap/substract the value then now nothing has dropped but yet you allow other things to drop which will definitely be in the wrong order

Reply:
Actually, the best solution is to only panic when the counter is too low; it should be perfectly fine to create and `mem::forget` a guard; the next guard will just set the counter to the value it was when they found it and reset stdio to the resource it was when they found it.

However, if the counter is too low then a guard created before this guard existed has been dropped before this guard has been dropped; the other guard assumed that this guard had been `mem::forget`ed, but it's now being dropped. This is when a panic occurs, and the counter should be left untouched.

I'll see if I can implement this.

Reply:
> because during unwinding all the other guards will simply do their regular destruction mechanism, avoiding double-panicking

that might not be true, some of the guards can be in different threads, and my concern isn't about double-panicing, that's fine, it's about having a wrong counter (because you already decremented it before panicking) which might seem like it works fine but will end up in the wrong result because another one panicked.

Reply:
A panic will only ever cause the counter to be too high, and that's never a problem since too high counters are treated the exact same as correct counters. As such everything will end up with the correct result as resetting the stdio resource can only occur when the counter moves down, so guard droppers, if run at all, will run in the correct order.